packages/response-ops/alerts_table/components/alerts_table.tsx

```diff
@@ -0,0 +1,616 @@
+  columns.some(({ id }) => id === ALERT_MAINTENANCE_WINDOW_IDS);
+
+const emptyRowSelection = new Map<number, RowSelectionState>();
+
+const initialBulkActionsState = {
+  rowSelection: emptyRowSelection,
+  isAllSelected: false,
+  areAllVisibleRowsSelected: false,
+  rowCount: 0,
+  updatedAt: Date.now(),
+};
+
+/**
+ * An `EuiDataGrid` abstraction to render alert documents
+ *
+ * It manages the paginated and cached fetching of alerts based on the
+ * provided `ruleTypeIds` and `consumers` (the final query can be refined
+ * through the `query` and `initialSort` props). The `id` prop is required in order
+ * to persist the table state in `localStorage`
+ *
+ * @example
+ * ```tsx
+ * <AlertsTable
+ *   id="my-alerts-table"
+ *   ruleTypeIds={ruleTypeIds}
+ *   consumers={consumers}
+ *   query={esQuery}
+ *   initialSort={defaultAlertsTableSort}
+ *   renderCellValue={CellValue}
+ *   renderActionsCell={ActionsCell}
+ *   services={{ ... }}
+ * />
+ * ```
+ */
+export const AlertsTable = memo(
+  forwardRef((props, ref) => {
+    return (
+      <QueryClientProvider client={alertsTableQueryClient} context={AlertsQueryContext}>
+        <ErrorBoundary fallback={ErrorFallback}>
+          <AlertsTableContent {...props} ref={ref} />
+        </ErrorBoundary>
+      </QueryClientProvider>
+    );
+  })
+  // Type cast to avoid losing the generic type
+) as typeof AlertsTableContent;
+
+(AlertsTable as FC).displayName = 'AlertsTable';
+
+const DEFAULT_LEADING_CONTROL_COLUMNS: EuiDataGridControlColumn[] = [];
+const DEFAULT_SORT: SortCombinations[] = [];
+const DEFAULT_COLUMNS: EuiDataGridColumn[] = [];
+
+const AlertsTableContent = typedForwardRef(
+  <AC extends AdditionalContext>(
+    {
+      id,
+      ruleTypeIds,
+      consumers,
+      query,
+      initialSort = DEFAULT_SORT,
+      initialPageSize = DEFAULT_ALERTS_PAGE_SIZE,
+      leadingControlColumns = DEFAULT_LEADING_CONTROL_COLUMNS,
+      trailingControlColumns,
+      rowHeightsOptions,
+      columns: initialColumns = defaultAlertsTableColumns,
+      gridStyle,
+      browserFields: propBrowserFields,
+      onUpdate,
+      onLoaded,
+      runtimeMappings,
+      showAlertStatusWithFlapping,
+      toolbarVisibility,
+      shouldHighlightRow,
+      dynamicRowHeight = false,
+      emptyStateHeight,
+      additionalContext,
+      renderCellValue,
+      renderCellPopover,
+      renderActionsCell,
+      renderFlyoutHeader,
+      renderFlyoutBody,
+      renderFlyoutFooter,
+      renderAdditionalToolbarControls: AdditionalToolbarControlsComponent,
+      lastReloadRequestTime,
+      services,
+      ...publicDataGridProps
+    }: AlertsTableProps<AC>,
+    ref: Ref<AlertsTableImperativeApi>
+  ) => {
+    // Memoized so that consumers can pass an inline object without causing re-renders
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const memoizedServices = useMemo(() => services, Object.values(services));
```

Comment:
I don't think this will work, as it returns a new array every time. why-did-you-render was flagging this as well https://github.com/welldone-software/why-did-you-render

Reply:
I tested this pretty thoroughly and it seems to work pretty well 🤔 The `Object.values` deps array is a bit cryptic probably... I can extract all the services to avoid any issues or improve legibility 🙂